File: python/callbackCrypto.py
import base64
import hashlib
import random
import string
from Crypto.Cipher import AES
import time


class CallbackCrypto:
    RANDOM_LENGTH = 16

    def __init__(self, encoding_aes_key, client_id):
        if not encoding_aes_key:
            raise ValueError("Encryption key cannot be empty")

        # 解码Base64密钥并验证长度
        self._aes_key = base64.b64decode(encoding_aes_key)
        if len(self._aes_key) != 16:
            raise ValueError(
                f"Invalid AES key length: expected 16 bytes, got {len(self._aes_key)} bytes"
            )

        self._client_id = client_id

    def get_encrypted_map(self, plaintext, timestamp=None):
        if not plaintext:
            raise ValueError("Plaintext cannot be empty")

        timestamp = timestamp or str(int(time.time() * 1000))
        nonce = Utils.get_random_str(CallbackCrypto.RANDOM_LENGTH)
        encrypt = self.encrypt(nonce, plaintext)
        signature = self.get_signature(timestamp, nonce, encrypt)

        return {
            "signature": signature,
            "encrypt": encrypt,
            "timestamp": timestamp,
            "nonce": nonce,
        }

    def get_decrypt_msg(self, msg_signature, timestamp, nonce, encrypt_msg):
        signature = self.get_signature(timestamp, nonce, encrypt_msg)
        if signature != msg_signature:
            raise ValueError("Signature mismatch")
        return self.decrypt(encrypt_msg)

    def encrypt(self, nonce, plaintext):
        # 构建数据结构: 随机串 + 内容长度(4字节) + 明文 + client_id
        random_bytes = nonce.encode("utf-8")
        plaintext_bytes = plaintext.encode("utf-8")
        length_bytes = Utils.int_to_bytes(len(plaintext_bytes))
        client_id_bytes = self._client_id.encode("utf-8")

        data = random_bytes + length_bytes + plaintext_bytes + client_id_bytes
        padded_data = PKCS7Padding.get_padding_bytes(data)

        # AES-CBC加密
        cipher = AES.new(self._aes_key, AES.MODE_CBC, self._aes_key[:16])
        encrypted = cipher.encrypt(padded_data)
        return base64.b64encode(encrypted).decode("utf-8")

    def decrypt(self, text):
        # 解密过程
        encrypted_data = base64.b64decode(text)
        decipher = AES.new(self._aes_key, AES.MODE_CBC, self._aes_key[:16])
        decrypted = decipher.decrypt(encrypted_data)

        # 去除填充
        unpadded = PKCS7Padding.remove_padding_bytes(decrypted)

        # 解析数据结构
        random_length = CallbackCrypto.RANDOM_LENGTH
        length_bytes = unpadded[random_length : random_length + 4]
        plaintext_length = Utils.bytes_to_int(length_bytes)

        plaintext = unpadded[
            random_length + 4 : random_length + 4 + plaintext_length
        ].decode("utf-8")
        msg_client_id = unpadded[random_length + 4 + plaintext_length :].decode("utf-8")

        if msg_client_id != self._client_id:
            raise ValueError("ClientID mismatch in decrypted message")

        return plaintext

    def get_signature(self, timestamp, nonce, encrypt):
        # 生成签名
        array = [self._client_id, timestamp, nonce, encrypt]
        array.sort()
        data = "".join(array)
        sha1 = hashlib.sha1()
        sha1.update(data.encode("ascii"))
        return sha1.hexdigest()

    def verify_signature(self, signature, timestamp, nonce, encrypt):
        return signature == self.get_signature(str(timestamp) , nonce, encrypt)


class Utils:
    @staticmethod
    def get_random_str(count):
        # 生成指定长度的随机字符串
        chars = string.ascii_letters + string.digits
        return "".join(random.choice(chars) for _ in range(count))

    @staticmethod
    def int_to_bytes(number):
        # 整数转4字节大端序
        return number.to_bytes(4, byteorder="big", signed=False)

    @staticmethod
    def bytes_to_int(byte_arr):
        # 4字节大端序转整数
        return int.from_bytes(byte_arr, byteorder="big", signed=False)


class PKCS7Padding:
    BLOCK_SIZE = 32

    @staticmethod
    def get_padding_bytes(data):
        # 计算填充长度
        amount_to_pad = PKCS7Padding.BLOCK_SIZE - (len(data) % PKCS7Padding.BLOCK_SIZE)
        if amount_to_pad == 0:
            amount_to_pad = PKCS7Padding.BLOCK_SIZE

        # 生成填充字节
        pad_byte = bytes([amount_to_pad])
        return data + pad_byte * amount_to_pad

    @staticmethod
    def remove_padding_bytes(decrypted):
        # 去除填充
        if not decrypted:
            return decrypted

        pad = decrypted[-1]
        if 1 <= pad <= PKCS7Padding.BLOCK_SIZE:
            return decrypted[:-pad]
        return decrypted


# 需要安装pycryptodome库
# 安装命令: pip install pycryptodome

File: php/CallbackCrypto.php
<?php

class CallbackCrypto {
    const RANDOM_LENGTH = 16;
    private $_aesKey;
    private $_clientId;

    public function __construct($encodingAesKey, $clientId) {
        if (empty($encodingAesKey)) {
            throw new Exception('Encryption key cannot be empty');
        }
        $this->_aesKey = base64_decode($encodingAesKey);
        if (strlen($this->_aesKey) !== 16) {
            throw new Exception("Invalid AES key length: expected 16 bytes, got " . strlen($this->_aesKey) . " bytes");
        }
        $this->_clientId = $clientId;
    }

    public function getEncryptedMap($plaintext, $timestamp = null) {
        if (empty($plaintext)) {
            throw new Exception('Plaintext cannot be empty');
        }
        $timestamp = $timestamp ?? time() * 1000; // 毫秒时间戳
        $nonce = Utils::getRandomStr(self::RANDOM_LENGTH);
        $encrypt = $this->encrypt($nonce, $plaintext);
        $signature = $this->getSignature((string)$timestamp, $nonce, $encrypt);

        return [
            'signature' => $signature,
            'encrypt' => $encrypt,
            'timestamp' => (string)$timestamp,
            'nonce' => $nonce
        ];
    }

    public function getDecryptMsg($msgSignature, $timestamp, $nonce, $encryptMsg) {
        $signature = $this->getSignature($timestamp, $nonce, $encryptMsg);
        if ($signature !== $msgSignature) {
            throw new Exception('Signature mismatch');
        }
        return $this->decrypt($encryptMsg);
    }

    private function encrypt($nonce, $plaintext) {
        $randomBytes = $nonce;
        $plainTextBytes = $plaintext;
        $lengthBytes = Utils::int2Bytes(strlen($plainTextBytes));
        $clientIdBytes = $this->_clientId;

        $data = $randomBytes . $lengthBytes . $plainTextBytes . $clientIdBytes;
        $paddedData = PKCS7Padding::getPaddingBytes($data);

        $iv = substr($this->_aesKey, 0, 16);
        $encrypted = openssl_encrypt(
            $paddedData,
            'aes-128-cbc',
            $this->_aesKey,
            OPENSSL_RAW_DATA | OPENSSL_NO_PADDING,
            $iv
        );

        return base64_encode($encrypted);
    }

    private function decrypt($text) {
        $cipherText = base64_decode($text);
        $iv = substr($this->_aesKey, 0, 16);
        $decrypted = openssl_decrypt(
            $cipherText,
            'aes-128-cbc',
            $this->_aesKey,
            OPENSSL_RAW_DATA | OPENSSL_NO_PADDING,
            $iv
        );

        $unpadded = PKCS7Padding::removePaddingBytes($decrypted);
        $plainTextLength = Utils::bytes2Int(substr($unpadded, self::RANDOM_LENGTH, 4));
        $plaintext = substr($unpadded, self::RANDOM_LENGTH + 4, $plainTextLength);
        $msgClientId = substr($unpadded, self::RANDOM_LENGTH + 4 + $plainTextLength);

        if ($msgClientId !== $this->_clientId) {
            throw new Exception('ClientID mismatch in decrypted message');
        }
        return $plaintext;
    }

    private function getSignature($timestamp, $nonce, $encrypt) {
        $array = [$this->_clientId, $timestamp, $nonce, $encrypt];
        sort($array, SORT_STRING);
        $data = implode('', $array);
        return sha1($data);
    }
}

class Utils {
    public static function getRandomStr($count) {
        $chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        $result = '';
        for ($i = 0; $i < $count; $i++) {
            $result .= $chars[mt_rand(0, strlen($chars) - 1)];
        }
        return $result;
    }

    public static function int2Bytes($number) {
        $buffer = pack('N', $number); // 32位大端字节序
        return $buffer;
    }

    public static function bytes2Int($byteArr) {
        $unpacked = unpack('N', $byteArr); // 32位大端字节序解析
        return $unpacked[1];
    }
}

class PKCS7Padding {
    const BLOCK_SIZE = 32;

    public static function getPaddingBytes($data) {
        $dataLen = strlen($data);
        $amountToPad = self::BLOCK_SIZE - ($dataLen % self::BLOCK_SIZE);
        if ($amountToPad === 0) {
            $amountToPad = self::BLOCK_SIZE;
        }
        $padChr = chr($amountToPad);
        return $data . str_repeat($padChr, $amountToPad);
    }

    public static function removePaddingBytes($decrypted) {
        $pad = ord(substr($decrypted, -1));
        if ($pad < 1 || $pad > self::BLOCK_SIZE) {
            return $decrypted;
        }
        return substr($decrypted, 0, strlen($decrypted) - $pad);
    }
}